### PR TITLE
[Bugfix] Fix UVA offload fallback copies

### PR DESCRIPTION
**Pull request by wgu9**
## Purpose

Fixes the non-UVA fallback path reported in #37883. The issue was closed automatically, but received new reports on August 10 and 11.

When UVA is unavailable or disabled, `UVAOffloader` copies offloaded module state to the accelerator for each forward call. The previous wrapper had several problems:

- `state_dict()` inside `forward()` is not traceable by Dynamo.
- Temporarily replacing `module.forward` lets exceptions and concurrent calls lose or bypass the wrapper.
- State captured before post-load processing can become stale when parameters, buffers, or child modules are replaced.
- Calling `functional_call` on the wrapped module dispatches module hooks a second time.
- CPU-to-device copies used `non_blocking=True` even when CPU memory was not pinned.

The wrapper now stays installed during every call. A private proxy invokes the saved original forward while `functional_call` supplies the current parameters and persistent buffers. Eager calls use a per-module reentrant lock because functional reparameterization is temporary. Dynamo capture bypasses the lock and traces the functional call. The copy mode now follows the actual pin-memory setting.

## Duplicate check

This does not duplicate the other open offloader changes:

- #43453 keeps externally cached quantization scales on the accelerator. It does not change fallback invocation or state resolution.
- #48468 wires the offloader into the V2 model runner. It does not change `UVAOffloader`.
- #41496 enables pinned memory on WSL2 so that WSL can use UVA. This PR fixes the fallback used when UVA remains unavailable or disabled.
- Closed #29054 addressed the `state_dict()` trace failure, but not exception recovery, replaced state, hook behavior, or concurrent eager calls.

## Validation

Human review and test:

```text
.venv/bin/python -m pytest -q tests/model_executor/test_uva_offloader.py
10 passed
```

Repository checks:

```text
pre-commit run --files tests/model_executor/test_uva_offloader.py vllm/model_executor/offloader/uva.py
all applicable hooks passed
```

RTX 5090 checks also passed for in-place fullgraph compilation, recursive calls, CUDA Graph capture and replay, non-default stream copies, and 200 compiled calls across eight threads. Parameters remained on CPU between calls.

A model evaluation was not run because this patch does not change model math or output selection. The focused suite compares outputs across eager and compiled calls and covers post-load state replacement.

## AI assistance

OpenAI Codex assisted with the implementation and test preparation. The human submitter reviewed every changed line and ran the focused test suite before the branch update.

### CODE REVIEW

**Comment by wgu9** [comment]
I rechecked this against the vLLM AGENTS.md guidance and still believe this is a small but meaningful runtime fix, not busywork.

Why it is worth keeping open:
- It targets an open user-visible failure path from #37883, specifically the non-UVA fallback wrapper behavior.
- It is not a duplicate of #43453: that PR handles cached quantization scale parameters on GPU, while this PR removes `state_dict()` from `forward()` for Dynamo/torch.compile compatibility and avoids non-blocking CPU->device copies when the offloader did not pin CPU memory.
- It is not a duplicate of merged #41496: that enables pinned memory for WSL2 so the UVA path can be used; this PR keeps the fallback path correct when UVA or pinned memory is unavailable or disabled.
- The diff is scoped to the offloader owner and has targeted tests for both invariants.

The current failing `pre-run-check` is the contributor gate, not a code failure: it requires a maintainer-applied `verified` or `ready` label, or 4+ merged vLLM PRs. Local validation remains:

- `python -m ruff check vllm/model_executor/offloader/uva.py tests/model_executor/test_uva_offloader.py`
- `python -m ruff format --check vllm/model_executor/offloader/uva.py tests/model_executor/test_uva_offloader.py`
- `python -m pytest -q --confcutdir=tests/model_executor tests/model_executor/test_uva_offloader.py`

Could a maintainer please add `verified`/`ready` if this looks worth CI review?

**Comment by wgu9** [comment]
Maintainer note: this PR is ready for review from my side. The only failing check is the repository pre-run gate for contributors without 4 merged PRs / without `ready` or `verified`; local targeted validation and duplicate/value re-evaluation are in the PR description.

**Comment by wgu9** [comment]
Friendly maintainer review ping. This focused UVA offload fallback fix is conflict-free from my side. The only failing pre-run check requires a maintainer-applied `ready` or `verified` label before full CI can run; please add the appropriate label if the change is ready for validation. I am happy to address any review feedback.

**Comment by wgu9** [comment]
Rebased this onto current main and expanded the fallback coverage. The wrapper no longer swaps `module.forward` during calls. It resolves current parameters and persistent buffers for each invocation, preserves module hook behavior, and serializes eager functional reparameterization.

Human review is complete. The focused test passed:

```text
.venv/bin/python -m pytest -q tests/model_executor/test_uva_offloader.py
10 passed
```

All applicable pre-commit hooks passed. RTX 5090 checks also passed for in-place fullgraph compilation, CUDA Graph capture and replay, non-default stream copies, recursive calls, and 200 compiled calls across eight threads.
